File: code/library/src/main/java/org/fbme/lib/iec61499/declarations/AlgorithmLanguage.java
package org.fbme.lib.iec61499.declarations;

import java.util.HashMap;
import java.util.Map;

public class AlgorithmLanguage<BodyT extends AlgorithmBody> {

    private final String myLanguageName;

    public AlgorithmLanguage(String languageName) {
        myLanguageName = languageName;
    }

    public String getLanguageName() {
        return myLanguageName;
    }

    @Override
    public String toString() {
        return String.format("AlgorithmLanguage{%s}", myLanguageName);
    }

    public static AlgorithmLanguage<AlgorithmBody.ST> ST = new AlgorithmLanguage<>("ST");

    private static final Map<String, AlgorithmLanguage<AlgorithmBody.Unknown>> ourUnknownLanguages = new HashMap<>();

    public static AlgorithmLanguage<AlgorithmBody.Unknown> unknown(String languageName) {
        return ourUnknownLanguages.computeIfAbsent(languageName, AlgorithmLanguage::new);
    }

    public static boolean isUnknown(AlgorithmLanguage<?> language) {
        return ourUnknownLanguages.get(language.getLanguageName()) == language;
    }
}

File: code/library/src/main/kotlin/org/fbme/lib/iec61499/declarations/AlgorithmLanguage.kt
package org.fbme.lib.iec61499.declarations

class AlgorithmLanguage<BodyT : AlgorithmBody>(val languageName: String) {
    override fun toString(): String {
        return String.format("AlgorithmLanguage{%s}", languageName)
    }

    companion object {
        @JvmField
        var ST = AlgorithmLanguage<AlgorithmBody.ST>("ST")
        private val ourUnknownLanguages: MutableMap<String, AlgorithmLanguage<AlgorithmBody.Unknown>> = HashMap()

        @JvmStatic
        fun unknown(languageName: String): AlgorithmLanguage<AlgorithmBody.Unknown> {
            return ourUnknownLanguages.computeIfAbsent(languageName) { AlgorithmLanguage(it) }
        }

        @JvmStatic
        fun isUnknown(language: AlgorithmLanguage<*>): Boolean {
            return ourUnknownLanguages[language.languageName] === language
        }
    }
}
